drivers/entropy/entropy_sam.c

```diff
@@ -53,20 +53,9 @@ static int entropy_sam_wait_ready(Trng * const trng, uint32_t flags)
 	int timeout = 1000000;
 
 	while (!_ready(trng)) {
-		if (timeout-- == 0) {
+		if ((timeout-- == 0) || ((flags & ENTROPY_BUSYWAIT) == 0U)) {
 			return -ETIMEDOUT;
 		}
-
-		if ((flags & ENTROPY_BUSYWAIT) == 0U) {
-			/* This internal function is used by both get_entropy,
-			 * and get_entropy_isr APIs. The later may call this
-			 * function with the ENTROPY_BUSYWAIT flag set. In
-			 * that case make no assumption that the kernel is
-			 * initialized when the function is called; so, just
-			 * do busy-wait for the random data to be ready.
-			 */
-			k_yield();
```

Comment:
Can you proposal be achieved by

```diff
- k_yield();
+ return -EAGAIN;
```

When it should return immediately, it could be more interesting return with another error like EAGAIN. I'm open to other possible error that could fit better in this situation too.

What do you think?

Reply:
`ETIMEDOUT` is for hw issue ? if not shouldn't both cases return same error ? `EAGAIN` is more common indeed.